### PR TITLE
Add fp16 support to collective functions

### PR DESCRIPTION
**Pull request by belldandyxtq**
This commit adds fp16 tests to collective functions.
The tests are also changed from unittest to plain tests to use the
parameterize.

Thank you for creating a pull request!

Please double-check the following.

- Read [our contribution guide](https://docs.chainer.org/en/stable/contribution.html).
  - Does your code conform to our coding guidelines?
  - Did you write sufficient test code?
  - Did you write sufficient documentation?
- Also, take a look at [our compatibility policy](https://docs.chainer.org/en/stable/compatibility.html).

### CODE REVIEW

**Comment by belldandyxtq** [comment]
pfnci, test this please

**Comment by belldandyxtq** [comment]
pfnci, test this please

**Comment by belldandyxtq** [comment]
pfnci, test this please

**Comment by belldandyxtq** [comment]
pfnci, test this please

**Comment by shu65** [comment]
I got SEGV when I performed chainermn unittests.
The failed test case is:
```
chainermn_tests/functions_tests/test_collective_communication.py::test_all_gather_cpu[param0]
```

**Comment by belldandyxtq** [comment]
pfnci, test this please

**Comment by belldandyxtq** [comment]
pfnci, test this please

File: chainermn/functions/collective_communication.py [hunk 1]
```diff
@@ -93,17 +160,38 @@ def __init__(self, comm, root):
...
+        # convert to float32 for communication
+        config_dtype = chainer.config.dtype
+        if np.float32 != config_dtype:
+            x = x.astype(np.float32)
         ys = self.comm.gather(x, self.root)
 
         if self.comm.rank == self.root:
+
+            # convert back
+            if np.float32 != config_dtype:
```
Comment (shu65) [inline chainermn/functions/collective_communication.py]: Same as above

File: chainermn/functions/collective_communication.py [hunk 2]
```diff
@@ -93,17 +160,38 @@ def __init__(self, comm, root):
...
+        config_dtype = chainer.config.dtype
+
+        # convert to float32 for communication
+        if self.comm.rank == self.root and np.float32 != config_dtype:
+            grad_outputs = tuple([item.astype(np.float32)
+                                  for item in grad_outputs])
+        ret = self.comm.scatter(grad_outputs, self.root),
+
+        # convert back
+        if np.float32 != config_dtype:
```
Comment (shu65) [inline chainermn/functions/collective_communication.py]: Same as above

File: chainermn/functions/collective_communication.py [hunk 3]
```diff
@@ -120,35 +208,57 @@ def __call__(self, *inputs):
...
+
+            # convert to float32 for communication
+            if np.float32 != config_dtype:
+                inputs = tuple([item.astype(np.float32) for item in inputs])
             y = self.comm.scatter(inputs, self.root)
         else:
             y = self.comm.scatter(None, self.root)
 
+        # convert back
+        if np.float32 != config_dtype:
```
Comment (shu65) [inline chainermn/functions/collective_communication.py]: Same as above

File: chainermn/functions/collective_communication.py [hunk 4]
```diff
@@ -120,35 +208,57 @@ def __call__(self, *inputs):
...
+            y = y.astype(config_dtype)
+
         return y,
 
     def backward(self, inputs, grad_outputs):
         xp = backend.get_array_module(*inputs)
         gy, = grad_outputs
+        config_dtype = chainer.config.dtype
```
Comment (shu65) [inline chainermn/functions/collective_communication.py]: Same as above

File: chainermn/functions/collective_communication.py [hunk 5]
```diff
@@ -120,35 +208,57 @@ def __call__(self, *inputs):
...
+
+        # convert to float32 for communication
+        if np.float32 != config_dtype:
```
Comment (shu65) [inline chainermn/functions/collective_communication.py]: Same as above

File: chainermn/functions/collective_communication.py [hunk 6]
```diff
@@ -120,35 +208,57 @@ def __call__(self, *inputs):
...
+            gy = gy.astype(np.float32)
+
         gxs = self.comm.gather(gy, self.root)
 
         if self.comm.rank == self.root:
+
+            # convert back
+            if np.float32 != config_dtype:
```
Comment (shu65) [inline chainermn/functions/collective_communication.py]: Same as above

File: chainermn/functions/collective_communication.py [hunk 7]
```diff
@@ -120,35 +208,57 @@ def __call__(self, *inputs):
...
+            dummy_var = chainer.Variable(
+                xp.array([], dtype=chainer.config.dtype))
             dummy_var.name = 'dummy_var'
             return super(Scatter, self).__call__(dummy_var)
 
         else:
             return super(Scatter, self).__call__(*inputs)
 
     def forward(self, inputs):
+        config_dtype = chainer.config.dtype
```
Comment (shu65) [inline chainermn/functions/collective_communication.py]: Same as above

File: chainermn/functions/collective_communication.py [hunk 8]
```diff
@@ -120,35 +208,57 @@ def __call__(self, *inputs):
...
         if self.comm.rank == self.root:
+
+            # convert to float32 for communication
+            if np.float32 != config_dtype:
```
Comment (shu65) [inline chainermn/functions/collective_communication.py]: Same as above

**Comment by belldandyxtq** [comment]
chainerci, test this please

**Comment by belldandyxtq** [comment]
chainerci, test this please

**Pull request merged**